segfault in jacobi_kernels test with Ginkgo 1.3.0 built with GCC 10.2.0

I'm seeing failing tests when building Ginkgo 1.3.0 with GCC 10.2.0 using [EasyBuild](https://easybuild.io), on various systems.

I've briefly discussed this with @upsj, who asked me to open an issue with all details included.

Failing test, when building with CMake 3.18.4 + `make` on top of CUDA 11.1.1, then running the tests with `make test`:

```
$ make test
...
The following tests FAILED:                                                                                                                                                                                     
    152 - omp/test/preconditioner/jacobi_kernels (SEGFAULT)
Errors while running CTest                                                                                                                                                                                  
make: *** [test] Error 8
```

* system info: CentOS 7.9, Intel Cascade Lake (Intel Xeon Gold 6242)
* compiler options: `-O2 -ftree-vectorize -march=native -fno-math-errno` (but I'm seeing the same failing test when trimming this down)
* corresponding [CMakeCache.txt](https://github.com/ginkgo-project/ginkgo/files/6246006/CMakeCache.txt)

A very similar problem happens on another system (RHEL 8.2, AMD Rome, AMD EPYC 7552), without CUDA, when using CMake 3.18.4 + Ninja 1.10.1, same compiler options:

```
$ ninja test
...
The following tests FAILED:
    116 - omp/test/preconditioner/jacobi_kernels (SEGFAULT)
    118 - omp/test/solver/bicg_kernels (Failed)
Errors while running CTest
```

* corresponding [CMakeCache.txt](https://github.com/ginkgo-project/ginkgo/files/6246108/CMakeCache.txt)
* corresponding [build.ninja](https://github.com/ginkgo-project/ginkgo/files/6246113/build.ninja.txt)

Some hopefully useful info collected with GDB:

```
$ gdb omp/test/preconditioner/jacobi_kernels
(gdb) run
...
[==========] Running 39 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 39 tests from Jacobi
[ RUN      ] Jacobi.OmpFindNaturalBlocksEquivalentToRef
[       OK ] Jacobi.OmpFindNaturalBlocksEquivalentToRef (54 ms)
[ RUN      ] Jacobi.OmpExecutesSupervariableAgglomerationEquivalentToRef
[       OK ] Jacobi.OmpExecutesSupervariableAgglomerationEquivalentToRef (12 ms)
[ RUN      ] Jacobi.OmpFindNaturalBlocksInLargeMatrixEquivalentToRef

Thread 13 "jacobi_kernels" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x155551f5e700 (LWP 1022726)]
0x00001555554ead13 in void gko::kernels::omp::jacobi::generate<double, int>(std::shared_ptr<gko::OmpExecutor const>, gko::matrix::Csr<double, int> const*, unsigned long, unsigned int, gko::detail::remove_complex_impl<double>::type, gko::preconditioner::block_interleaved_storage_scheme<int> const&, gko::Array<gko::detail::remove_complex_impl<double>::type>&, gko::Array<gko::precision_reduction>&, gko::Array<int> const&, gko::Array<double>&) [clone ._omp_fn.0] () from /tmp/slurm.50117758.0/tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/omp/libginkgo_omp.so.1.3.0
(gdb) bt
#0  0x00001555554ead13 in void gko::kernels::omp::jacobi::generate<double, int>(std::shared_ptr<gko::OmpExecutor const>, gko::matrix::Csr<double, int> const*, unsigned long, unsigned int, gko::detail::remove_complex_impl<double>::type, gko::preconditioner::block_interleaved_storage_scheme<int> const&, gko::Array<gko::detail::remove_complex_impl<double>::type>&, gko::Array<gko::precision_reduction>&, gko::Array<int> const&, gko::Array<double>&) [clone ._omp_fn.0] () from /tmp/slurm.50117758.0/tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/omp/libginkgo_omp.so.1.3.0
#1  0x00001555539e3036 in gomp_thread_start (xdata=<optimized out>) at ../../../libgomp/team.c:123
#2  0x00001555537b12de in start_thread () from /lib64/libpthread.so.0
#3  0x0000155553b04e83 in clone () from /lib64/libc.so.6
```

@upsj I tried doing a static build, but I'm running into some trouble there, linking errors like `cannot find -lgcc_s` because of missing static libraries in the GCC installation (which I'm not sure is easy to fix).

I do have a little bit of perhaps useful information though.

* All 39 tests when running `omp/test/preconditioner/jacobi_kernels` pass after setting `$OMP_NUM_THREADS` to `1`
* At one point when using `OMP_NUM_THREADS=2`, I got this:
  ```
  ...
  [ RUN      ] Jacobi.OmpApplyEquivalentToRefWithDifferentBlockSize
  [       OK ] Jacobi.OmpApplyEquivalentToRefWithDifferentBlockSize (7 ms)
  [ RUN      ] Jacobi.OmpApplyEquivalentToRef
  corrupted double-linked list
  Aborted (core dumped)
  ```
  That's not reproducible though, it often just crashes with a segfault instead.

* When running under Valgrind, the problem does not occur.
* When using `valgrind --leak-check=full`, I get this (not sure if that's useful at all):
```
==3508653== 304 bytes in 1 blocks are possibly lost in loss record 4 of 5
==3508653==    at 0x4036B35: calloc (vg_replace_malloc.c:760)
==3508653==    by 0x4012341: allocate_dtv (in /usr/lib64/ld-2.28.so)
==3508653==    by 0x4012CD1: _dl_allocate_tls (in /usr/lib64/ld-2.28.so)
==3508653==    by 0x62F7F32: pthread_create@@GLIBC_2.2.5 (in /usr/lib64/libpthread-2.28.so)
==3508653==    by 0x62C964A: gomp_team_start (team.c:839)
==3508653==    by 0x62C16FC: GOMP_parallel (parallel.c:169)
==3508653==    by 0x417BDC7: void gko::kernels::omp::jacobi::generate<double, int>(std::shared_ptr<gko::OmpExecutor const>, gko::matrix::Csr<double, int> const*, unsigned long, unsigned int, gko::detail::remove_complex_impl<double>::type, gko::preconditioner::block_interleaved_storage_scheme<int> const&, gko::Array<gko::detail::remove_complex_impl<double>::type>&, gko::Array<gko::precision_reduction>&, gko::Array<int> const&, gko::Array<double>&) (jacobi_kernels.cpp:374)
==3508653==    by 0x53E5801: call<0, 1, 2, 3, 4, 5, 6, 7, 8> (jacobi.cpp:57)
==3508653==    by 0x53E5801: gko::preconditioner::jacobi::generate_operation<gko::matrix::Csr<double, int> const*, unsigned long&, unsigned int&, double&, gko::preconditioner::block_interleaved_storage_scheme<int>&, gko::Array<double>&, gko::Array<gko::precision_reduction>&, gko::Array<int>&, gko::Array<double>&>::run(std::shared_ptr<gko::OmpExecutor const>) const (jacobi.cpp:57)
==3508653==    by 0x4089386: gko::detail::ExecutorBase<gko::OmpExecutor>::run(gko::Operation const&) const (executor.hpp:762)
==3508653==    by 0x5406C47: gko::preconditioner::Jacobi<double, int>::generate(gko::LinOp const*) (jacobi.cpp:231)
==3508653==    by 0x43E80D: gko::preconditioner::Jacobi<double, int>::Jacobi(gko::preconditioner::Jacobi<double, int>::Factory const*, std::shared_ptr<gko::LinOp const>) (jacobi.hpp:512)
==3508653==    by 0x43ED8A: generate_impl (abstract_factory.hpp:246)
==3508653==    by 0x43ED8A: std::unique_ptr<gko::preconditioner::Jacobi<double, int>, std::default_delete<gko::preconditioner::Jacobi<double, int> > > gko::EnableDefaultFactory<gko::preconditioner::Jacobi<double, int>::Factory, gko::preconditioner::Jacobi<double, int>, gko::preconditioner::Jacobi<double, int>::parameters_type, gko::LinOpFactory>::generate<std::shared_ptr<gko::matrix::Csr<double, int> >&>(std::shared_ptr<gko::matrix::Csr<double, int> >&) const (abstract_factory.hpp:165)
```
* Same problem occurs when building with `-O0 -g`. Here's the GDB backtrace I get with that build:
```
Thread 5 "jacobi_kernels" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x155551b20700 (LWP 3517893)]
0x000000000049a27d in gko::log::Logger::on<2ul, gko::Executor const*, unsigned long> (this=0x4000100010007)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/include/ginkgo/core/log/logger.hpp:168
168         GKO_LOGGER_REGISTER_EVENT(2, free_started, const Executor *exec,
(gdb) bt
#0  0x000000000049a27d in gko::log::Logger::on<2ul, gko::Executor const*, unsigned long> (this=0x4000100010007)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/include/ginkgo/core/log/logger.hpp:168
#1  0x0000000000494bde in gko::log::EnableLogging<gko::Executor, gko::log::Loggable>::log<2ul, gko::Executor const*, unsigned long> (this=0x56c6a0)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/include/ginkgo/core/log/logger.hpp:547
#2  0x0000000000491ca1 in gko::Executor::free (this=0x56c6a0, ptr=0x15551c001420) at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/include/ginkgo/core/base/executor.hpp:507
#3  0x00000000004bcc20 in gko::executor_deleter<double []>::operator() (this=0x15551c000ea0, ptr=0x15551c001420)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/include/ginkgo/core/base/executor.hpp:744
#4  0x00000000004bb1b0 in std::__invoke_impl<void, gko::executor_deleter<double []>&, double*> (__f=...)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/invoke.h:60
#5  0x00000000004b8036 in std::__invoke_r<void, gko::executor_deleter<double []>&, double*> (__fn=...)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/invoke.h:153
#6  0x00000000004b44bc in std::_Function_handler<void (double*), gko::executor_deleter<double []> >::_M_invoke(std::_Any_data const&, double*&&) (__functor=..., __args#0=@0x155551b1f950: 0x15551c001420)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/std_function.h:291
#7  0x00000000004acda9 in std::function<void (double*)>::operator()(double*) const (this=0x15551c000cf8, __args#0=0x15551c001420)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/std_function.h:622
#8  0x00000000004a45bf in std::unique_ptr<double [], std::function<void (double*)> >::~unique_ptr() (this=0x15551c000cf8, __in_chrg=<optimized out>)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/unique_ptr.h:612
#9  0x000000000049c4e8 in gko::Array<double>::~Array (this=0x15551c000cf0, __in_chrg=<optimized out>)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/include/ginkgo/core/base/array.hpp:84
#10 0x000015555337e1ce in std::_Destroy<gko::Array<double> > (__pointer=0x15551c000cf0)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/stl_construct.h:140
#11 0x000015555337ce7c in std::allocator_traits<gko::ExecutorAllocator<gko::Array<double> > >::_S_destroy<gko::ExecutorAllocator<gko::Array<double> >, gko::Array<double> > (__p=0x15551c000cf0)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/alloc_traits.h:274
#12 0x000015555337afe8 in std::allocator_traits<gko::ExecutorAllocator<gko::Array<double> > >::destroy<gko::Array<double> > (__a=..., __p=0x15551c000cf0)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/alloc_traits.h:374
#13 0x0000155553378d04 in std::_Destroy<gko::Array<double>*, gko::ExecutorAllocator<gko::Array<double> > > (warning: (Internal error: pc 0x15555338145b in read in psymtab, but not in symtab.)

__first=0x15551c000cf0, __last=0x15551c000df0, __alloc=...)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/alloc_traits.h:728
#14 0x00001555533704be in std::vector<gko::Array<double>, gko::ExecutorAllocator<gko::Array<double> > >::~vector (warning: (Internal error: pc 0x15555338145b in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x155553381102 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x15555338145b in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x15555338145b in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x15555338145b in read in psymtab, but not in symtab.)

this=0x155551b1fb40, __in_chrg=<optimized out>)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/stl_vector.h:680
warning: (Internal error: pc 0x15555338145b in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x15555338145b in read in psymtab, but not in symtab.)

#15 0x000015555338145c in gko::kernels::omp::jacobi::_ZN3gko7kernels3omp6jacobi8generateIdiEEvSt10shared_ptrIKNS_11OmpExecutorEEPKNS_6matrix3CsrIT_T0_EEmjNS_6detail19remove_complex_implISA_E4typeERKNS_14preconditioner32block_interleaved_storage_schemeISB_EERNS_5ArrayISI_EERNSO_INS_19precision_reductionEEERKNSO_ISB_EERNSO_ISA_EE._omp_fn.0(void) (warning: (Internal error: pc 0x15555338145b in read in psymtab, but not in symtab.)

)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/omp/preconditioner/jacobi_kernels.cpp:376
#16 0x0000155555385036 in gomp_thread_start (xdata=<optimized out>) at ../../../libgomp/team.c:123
#17 0x00001555523622de in start_thread () from /lib64/libpthread.so.0
#18 0x0000155552675e83 in clone () from /lib64/libc.so.6
```

And with `OMP_NUM_THREADS=2`, here's the GDB backtrace with the `-O0 -g` build:
```
...
[ RUN      ] Jacobi.OmpConjTransposedPreconditionerEquivalentToRefWithMPW
[       OK ] Jacobi.OmpConjTransposedPreconditionerEquivalentToRefWithMPW (31 ms)
[ RUN      ] Jacobi.OmpApplyEquivalentToRefWithBlockSize32
corrupted double-linked list

Thread 1 "jacobi_kernels" received signal SIGABRT, Aborted.
0x00001555525b170f in raise () from /lib64/libc.so.6
(gdb) bt
#0  0x00001555525b170f in raise () from /lib64/libc.so.6
#1  0x000015555259bb25 in abort () from /lib64/libc.so.6
#2  0x00001555525f4897 in __libc_message () from /lib64/libc.so.6
#3  0x00001555525fafdc in malloc_printerr () from /lib64/libc.so.6
#4  0x00001555525fb81c in unlink_chunk.isra () from /lib64/libc.so.6
#5  0x00001555525fb953 in malloc_consolidate () from /lib64/libc.so.6
#6  0x00001555525fdd58 in _int_malloc () from /lib64/libc.so.6
#7  0x00001555525ff662 in malloc () from /lib64/libc.so.6
#8  0x0000155552d77755 in operator new (sz=2048) at ../../../../libstdc++-v3/libsupc++/new_op.cc:50
#9  0x00000000004b5786 in __gnu_cxx::new_allocator<gko::matrix_data<double, int>::nonzero_type>::allocate (this=0x7fffffff1dc0, __n=128)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/ext/new_allocator.h:115
#10 0x00000000004b046c in std::allocator_traits<std::allocator<gko::matrix_data<double, int>::nonzero_type> >::allocate (__a=..., __n=128)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/alloc_traits.h:460
#11 0x00000000004a86c8 in std::_Vector_base<gko::matrix_data<double, int>::nonzero_type, std::allocator<gko::matrix_data<double, int>::nonzero_type> >::_M_allocate (this=0x7fffffff1dc0, __n=128)
    at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/stl_vector.h:346
#12 0x00000000004a4150 in std::vector<gko::matrix_data<double, int>::nonzero_type, std::allocator<gko::matrix_data<double, int>::nonzero_type> >::_M_realloc_insert<unsigned long&, unsigned long&, double>
    (this=0x7fffffff1dc0, __position=...) at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/vector.tcc:440
#13 0x000000000049bcfe in std::vector<gko::matrix_data<double, int>::nonzero_type, std::allocator<gko::matrix_data<double, int>::nonzero_type> >::emplace_back<unsigned long&, unsigned long&, double> (
    this=0x7fffffff1dc0) at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/vector.tcc:121
#14 0x0000000000496490 in gko::test::generate_random_matrix<gko::matrix::Csr<double, int>, std::uniform_int_distribution<int>, std::normal_distribution<double>, std::discard_block_engine<std::subtract_with_carry_engine<unsigned long, 48ul, 5ul, 12ul>, 389ul, 11ul>&>(unsigned long, unsigned long, std::uniform_int_distribution<int>&&, std::normal_distribution<double>&&, std::discard_block_engine<std::subtract_with_carry_engine<unsigned long, 48ul, 5ul, 12ul>, 389ul, 11ul>&, std::shared_ptr<gko::Executor const>)::{lambda(unsigned long)#1}::operator()(unsigned long) const (this=0x7fffffff1d30, col=52)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/core/test/utils/matrix_generator.hpp:118
#15 0x000000000049bd54 in std::for_each<__gnu_cxx::__normal_iterator<unsigned long*, std::vector<unsigned long, std::allocator<unsigned long> > >, gko::test::generate_random_matrix<gko::matrix::Csr<double, int>, std::uniform_int_distribution<int>, std::normal_distribution<double>, std::discard_block_engine<std::subtract_with_carry_engine<unsigned long, 48ul, 5ul, 12ul>, 389ul, 11ul>&>(unsigned long, unsigned long, std::uniform_int_distribution<int>&&, std::normal_distribution<double>&&, std::discard_block_engine<std::subtract_with_carry_engine<unsigned long, 48ul, 5ul, 12ul>, 389ul, 11ul>&, std::shared_ptr<gko::Executor const>)::{lambda(unsigned long)#1}>(__gnu_cxx::__normal_iterator<unsigned long*, std::vector<unsigned long, std::allocator<unsigned long> > >, __gnu_cxx::__normal_iterator<unsigned long*, std::vector<unsigned long, std::allocator<unsigned long> > >, gko::test::generate_random_matrix<gko::matrix::Csr<double, int>, std::uniform_int_distribution<int>, std::normal_distribution<double>, std::discard_block_engine<std::subtract_with_carry_engine<unsigned long, 48ul, 5ul, 12ul>, 389ul, 11ul>&>(unsigned long, unsigned long, std::uniform_int_distribution<int>&&, std::normal_distribution<double>&&, std::discard_block_engine<std::subtract_with_carry_engine<unsigned long, 48ul, 5ul, 12ul>, 389ul, 11ul>&, std::shared_ptr<gko::Executor const>)::{lambda(unsigned long)#1}) (__first=..., __last=...,
    __f=...) at /arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild_tests/RHEL8/zen2-ib/software/GCCcore/10.2.0/include/c++/10.2.0/bits/stl_algo.h:3839
#16 0x00000000004966d7 in gko::test::generate_random_matrix<gko::matrix::Csr<double, int>, std::uniform_int_distribution<int>, std::normal_distribution<double>, std::discard_block_engine<std::subtract_with_carry_engine<unsigned long, 48ul, 5ul, 12ul>, 389ul, 11ul>&> (num_rows=128, num_cols=128, nonzero_dist=..., value_dist=..., engine=..., exec=...)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/core/test/utils/matrix_generator.hpp:116
#17 0x000000000047bf3c in (anonymous namespace)::Jacobi::initialize_data (this=0x56bf60, block_pointers=..., block_precisions=..., condition_numbers=..., max_block_size=32, min_nnz=100, max_nnz=111,
    num_rhs=1, accuracy=0.10000000000000001) at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/omp/test/preconditioner/jacobi_kernels.cpp:84
#18 0x0000000000480908 in (anonymous namespace)::Jacobi_OmpApplyEquivalentToRefWithBlockSize32_Test::TestBody (this=0x56bf60)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/ginkgo-1.3.0/omp/test/preconditioner/jacobi_kernels.cpp:357
#19 0x00000000004f71b4 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x56bf60, method=&virtual testing::Test::TestBody(), location=0x513163 "the test body")
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/src/gtest.cc:2437
#20 0x00000000004f1941 in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x56bf60, method=&virtual testing::Test::TestBody(), location=0x513163 "the test body")
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/src/gtest.cc:2473
#21 0x00000000004d3c56 in testing::Test::Run (this=0x56bf60) at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/src/gtest.cc:2511
#22 0x00000000004d450c in testing::TestInfo::Run (this=0x56d750) at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/src/gtest.cc:2687
#23 0x00000000004d4b5d in testing::TestCase::Run (this=0x56c530) at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/src/gtest.cc:2805
#24 0x00000000004de874 in testing::internal::UnitTestImpl::RunAllTests (this=0x56c2a0)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/src/gtest.cc:5131
--Type <RET> for more, q to quit, c to continue without paging--
#25 0x00000000004f830d in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x56c2a0,
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x4de5ce <testing::internal::UnitTestImpl::RunAllTests()>,
    location=0x513b58 "auxiliary test code (environments or event listeners)") at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/src/gtest.cc:2437
#26 0x00000000004f2837 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x56c2a0,
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x4de5ce <testing::internal::UnitTestImpl::RunAllTests()>,
    location=0x513b58 "auxiliary test code (environments or event listeners)") at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/src/gtest.cc:2473
#27 0x00000000004dd548 in testing::UnitTest::Run (this=0x559f30 <testing::UnitTest::GetInstance()::instance>)
    at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/src/gtest.cc:4740
#28 0x00000000004ccac6 in RUN_ALL_TESTS () at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/include/gtest/gtest.h:2333
#29 0x00000000004cca54 in main (argc=1, argv=0x7fffffff2ad8) at /tmp/vsc40023/easybuild_build/Ginkgo/1.3.0/GCC-10.2.0/easybuild_obj/third_party/gtest/src/googletest/src/gtest_main.cc:36
```

--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -437,7 +437,7 @@ build/cuda101/clang/all/release/static:
     CUDA_ARCH: 35
 
 # clang-cuda with cuda 10.1 and friends
-build/clang-cuda101/gcc/all/release/shared:
+build/clang-cuda101/gcc/cuda/release/shared:
   <<: *default_build
   extends:
     - .quick_test_condition
@@ -447,7 +447,7 @@ build/clang-cuda101/gcc/all/release/shared:
     CUDA_COMPILER: "clang++"
     BUILD_OMP: "ON"
     BUILD_CUDA: "ON"
-    BUILD_HIP: "ON"
+    BUILD_HIP: "OFF"
     BUILD_TYPE: "Release"
     CUDA_ARCH: 35
 

--- a/cmake/GinkgoConfig.cmake.in
+++ b/cmake/GinkgoConfig.cmake.in
@@ -170,8 +170,8 @@ if(GINKGO_HAVE_HWLOC)
     find_package(HWLOC REQUIRED)
 endif()
 
-# HIP depends on Threads::Threads in some circumstances, but doesn't find it
-if (GINKGO_BUILD_HIP)
+# HIP and OpenMP depend on Threads::Threads in some circumstances, but don't find it
+if (GINKGO_BUILD_HIP OR GINKGO_BUILD_OMP)
     find_package(Threads REQUIRED)
 endif()
 

--- a/omp/CMakeLists.txt
+++ b/omp/CMakeLists.txt
@@ -44,6 +44,17 @@ target_sources(ginkgo_omp
 
 ginkgo_compile_features(ginkgo_omp)
 
+# The dependency to OpenMP's potential pthread dependency needs to be public
+# because well, it's a long story:
+# std::shared_ptr decides whether we need to increment/decrement the reference
+# counters atomically by calling the pthread function __gthread_active_p, i.e.
+# checking whether there are multiple active threads. Linking against OpenMP
+# also links against pthread on Linux. Due to issues with static initialization
+# order for pthread_key_create, it may happen that ginkgo_omp assumes that a
+# piece of code is being executed sequentially, causing race conditions in the
+# reference counters.
+find_package(Threads REQUIRED)
+target_link_libraries(ginkgo_omp PUBLIC Threads::Threads)
 target_link_libraries(ginkgo_omp PRIVATE "${OpenMP_CXX_LIBRARIES}")
 target_include_directories(ginkgo_omp PRIVATE "${OpenMP_CXX_INCLUDE_DIRS}")
 # We first separate the arguments, otherwise, the target_compile_options adds it as a string

--- a/omp/matrix/csr_kernels.cpp
+++ b/omp/matrix/csr_kernels.cpp
@@ -34,6 +34,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
 #include <algorithm>
+#include <limits>
 #include <numeric>
 #include <utility>
 
@@ -51,6 +52,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "core/base/allocator.hpp"
 #include "core/base/iterator_factory.hpp"
+#include "core/base/utils.hpp"
 #include "core/components/prefix_sum.hpp"
 #include "core/matrix/csr_builder.hpp"
 #include "omp/components/csr_spgeam.hpp"
@@ -130,79 +132,131 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_CSR_ADVANCED_SPMV_KERNEL);
 
 
-template <typename ValueType, typename IndexType>
-void spgemm_insert_row(unordered_set<IndexType> &cols,
-                       const matrix::Csr<ValueType, IndexType> *c,
-                       size_type row)
-{
-    auto row_ptrs = c->get_const_row_ptrs();
-    auto col_idxs = c->get_const_col_idxs();
-    cols.insert(col_idxs + row_ptrs[row], col_idxs + row_ptrs[row + 1]);
-}
+namespace {
 
 
 template <typename ValueType, typename IndexType>
-void spgemm_insert_row2(unordered_set<IndexType> &cols,
-                        const matrix::Csr<ValueType, IndexType> *a,
-                        const matrix::Csr<ValueType, IndexType> *b,
-                        size_type row)
+struct col_heap_element {
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using matrix_type = matrix::Csr<ValueType, IndexType>;
+
+    IndexType idx;
+    IndexType end;
+    IndexType col;
+
+    ValueType val() const { return zero<ValueType>(); }
+
+    col_heap_element(IndexType idx, IndexType end, IndexType col, ValueType)
+        : idx{idx}, end{end}, col{col}
+    {}
+};
+
+
+template <typename ValueType, typename IndexType>
+struct val_heap_element {
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using matrix_type = matrix::Csr<ValueType, IndexType>;
+
+    IndexType idx;
+    IndexType end;
+    IndexType col;
+    ValueType val_;
+
+    ValueType val() const { return val_; }
+
+    val_heap_element(IndexType idx, IndexType end, IndexType col, ValueType val)
+        : idx{idx}, end{end}, col{col}, val_{val}
+    {}
+};
+
+
+template <typename HeapElement>
+void sift_down(HeapElement *heap, typename HeapElement::index_type idx,
+               typename HeapElement::index_type size)
 {
-    auto a_row_ptrs = a->get_const_row_ptrs();
-    auto a_col_idxs = a->get_const_col_idxs();
-    auto b_row_ptrs = b->get_const_row_ptrs();
-    auto b_col_idxs = b->get_const_col_idxs();
-    for (size_type a_nz = a_row_ptrs[row];
-         a_nz < size_type(a_row_ptrs[row + 1]); ++a_nz) {
-        auto a_col = a_col_idxs[a_nz];
-        auto b_row = a_col;
-        cols.insert(b_col_idxs + b_row_ptrs[b_row],
-                    b_col_idxs + b_row_ptrs[b_row + 1]);
+    auto curcol = heap[idx].col;
+    while (idx * 2 + 1 < size) {
+        auto lchild = idx * 2 + 1;
+        auto rchild = min(lchild + 1, size - 1);
+        auto lcol = heap[lchild].col;
+        auto rcol = heap[rchild].col;
+        auto mincol = min(lcol, rcol);
+        if (mincol >= curcol) {
+            break;
+        }
+        auto minchild = lcol == mincol ? lchild : rchild;
+        std::swap(heap[minchild], heap[idx]);
+        idx = minchild;
     }
 }
 
 
-template <typename ValueType, typename IndexType>
-void spgemm_accumulate_row(map<IndexType, ValueType> &cols,
-                           const matrix::Csr<ValueType, IndexType> *c,
-                           ValueType scale, size_type row)
-{
-    auto row_ptrs = c->get_const_row_ptrs();
-    auto col_idxs = c->get_const_col_idxs();
-    auto vals = c->get_const_values();
-    for (size_type c_nz = row_ptrs[row]; c_nz < size_type(row_ptrs[row + 1]);
-         ++c_nz) {
-        auto c_col = col_idxs[c_nz];
-        auto c_val = vals[c_nz];
-        cols[c_col] += scale * c_val;
-    }
-}
-
-
-template <typename ValueType, typename IndexType>
-void spgemm_accumulate_row2(map<IndexType, ValueType> &cols,
-                            const matrix::Csr<ValueType, IndexType> *a,
-                            const matrix::Csr<ValueType, IndexType> *b,
-                            ValueType scale, size_type row)
+template <typename HeapElement, typename InitCallback, typename StepCallback,
+          typename ColCallback>
+auto spgemm_multiway_merge(size_type row,
+                           const typename HeapElement::matrix_type *a,
+                           const typename HeapElement::matrix_type *b,
+                           HeapElement *heap, InitCallback init_cb,
+                           StepCallback step_cb, ColCallback col_cb)
+    -> decltype(init_cb(0))
 {
     auto a_row_ptrs = a->get_const_row_ptrs();
-    auto a_col_idxs = a->get_const_col_idxs();
+    auto a_cols = a->get_const_col_idxs();
     auto a_vals = a->get_const_values();
     auto b_row_ptrs = b->get_const_row_ptrs();
-    auto b_col_idxs = b->get_const_col_idxs();
+    auto b_cols = b->get_const_col_idxs();
     auto b_vals = b->get_const_values();
-    for (size_type a_nz = a_row_ptrs[row];
-         a_nz < size_type(a_row_ptrs[row + 1]); ++a_nz) {
-        auto a_col = a_col_idxs[a_nz];
-        auto a_val = a_vals[a_nz];
-        auto b_row = a_col;
-        for (size_type b_nz = b_row_ptrs[b_row];
-             b_nz < size_type(b_row_ptrs[b_row + 1]); ++b_nz) {
-            auto b_col = b_col_idxs[b_nz];
-            auto b_val = b_vals[b_nz];
-            cols[b_col] += scale * a_val * b_val;
+    auto a_begin = a_row_ptrs[row];
+    auto a_end = a_row_ptrs[row + 1];
+
+    using index_type = typename HeapElement::index_type;
+    constexpr auto sentinel = std::numeric_limits<index_type>::max();
+
+    auto state = init_cb(row);
+
+    // initialize the heap
+    for (auto a_nz = a_begin; a_nz < a_end; ++a_nz) {
+        auto b_row = a_cols[a_nz];
+        auto b_begin = b_row_ptrs[b_row];
+        auto b_end = b_row_ptrs[b_row + 1];
+        heap[a_nz] = {b_begin, b_end,
+                      checked_load(b_cols, b_begin, b_end, sentinel),
+                      a_vals[a_nz]};
+    }
+
+    if (a_begin != a_end) {
+        // make heap:
+        auto a_size = a_end - a_begin;
+        for (auto i = (a_size - 2) / 2; i >= 0; --i) {
+            sift_down(heap + a_begin, i, a_size);
+        }
+        auto &top = heap[a_begin];
+        auto &bot = heap[a_end - 1];
+        auto col = top.col;
+
+        while (top.col != sentinel) {
+            step_cb(b_vals[top.idx] * top.val(), top.col, state);
+            // move to the next element
+            top.idx++;
+            top.col = checked_load(b_cols, top.idx, top.end, sentinel);
+            // restore heap property
+            // pop_heap swaps top and bot, we need to prevent that
+            // so that we do a simple sift_down instead
+            sift_down(heap + a_begin, index_type{}, a_size);
+            if (top.col != col) {
+                col_cb(col, state);
+            }
+            col = top.col;
         }
     }
+
+    return state;
 }
+
+
+}  // namespace
 
 
 template <typename ValueType, typename IndexType>
@@ -212,17 +266,28 @@ void spgemm(std::shared_ptr<const OmpExecutor> exec,
             matrix::Csr<ValueType, IndexType> *c)
 {
     auto num_rows = a->get_size()[0];
-
-    // first sweep: count nnz for each row
     auto c_row_ptrs = c->get_row_ptrs();
 
-    unordered_set<IndexType> local_col_idxs(exec);
-#pragma omp parallel for firstprivate(local_col_idxs)
+    Array<col_heap_element<ValueType, IndexType>> col_heap_array(
+        exec, a->get_num_stored_elements());
+
+    auto col_heap = col_heap_array.get_data();
+
+    // first sweep: count nnz for each row
+#pragma omp parallel for
     for (size_type a_row = 0; a_row < num_rows; ++a_row) {
-        local_col_idxs.clear();
-        spgemm_insert_row2(local_col_idxs, a, b, a_row);
-        c_row_ptrs[a_row] = local_col_idxs.size();
+        c_row_ptrs[a_row] = spgemm_multiway_merge(
+            a_row, a, b, col_heap, [](size_type) { return IndexType{}; },
+            [](ValueType, IndexType, IndexType &) {},
+            [](IndexType, IndexType &nnz) { nnz++; });
     }
+
+    col_heap_array.clear();
+
+    Array<val_heap_element<ValueType, IndexType>> heap_array(
+        exec, a->get_num_stored_elements());
+
+    auto heap = heap_array.get_data();
 
     // build row pointers
     components::prefix_sum(exec, c_row_ptrs, num_rows + 1);
@@ -237,18 +302,21 @@ void spgemm(std::shared_ptr<const OmpExecutor> exec,
     auto c_col_idxs = c_col_idxs_array.get_data();
     auto c_vals = c_vals_array.get_data();
 
-    map<IndexType, ValueType> local_row_nzs(exec);
-#pragma omp parallel for firstprivate(local_row_nzs)
+#pragma omp parallel for
     for (size_type a_row = 0; a_row < num_rows; ++a_row) {
-        local_row_nzs.clear();
-        spgemm_accumulate_row2(local_row_nzs, a, b, one<ValueType>(), a_row);
-        // store result
-        auto c_nz = c_row_ptrs[a_row];
-        for (auto pair : local_row_nzs) {
-            c_col_idxs[c_nz] = pair.first;
-            c_vals[c_nz] = pair.second;
-            ++c_nz;
-        }
+        spgemm_multiway_merge(
+            a_row, a, b, heap,
+            [&](size_type row) {
+                return std::make_pair(zero<ValueType>(), c_row_ptrs[row]);
+            },
+            [](ValueType val, IndexType,
+               std::pair<ValueType, IndexType> &state) { state.first += val; },
+            [&](IndexType col, std::pair<ValueType, IndexType> &state) {
+                c_col_idxs[state.second] = col;
+                c_vals[state.second] = state.first;
+                state.first = zero<ValueType>();
+                state.second++;
+            });
     }
 }
 
@@ -267,17 +335,41 @@ void advanced_spgemm(std::shared_ptr<const OmpExecutor> exec,
     auto num_rows = a->get_size()[0];
     auto valpha = alpha->at(0, 0);
     auto vbeta = beta->at(0, 0);
+    constexpr auto sentinel = std::numeric_limits<IndexType>::max();
 
     // first sweep: count nnz for each row
     auto c_row_ptrs = c->get_row_ptrs();
+    auto d_row_ptrs = d->get_const_row_ptrs();
+    auto d_cols = d->get_const_col_idxs();
+    auto d_vals = d->get_const_values();
 
-    unordered_set<IndexType> local_col_idxs(exec);
-#pragma omp parallel for firstprivate(local_col_idxs)
+    Array<val_heap_element<ValueType, IndexType>> heap_array(
+        exec, a->get_num_stored_elements());
+
+    auto heap = heap_array.get_data();
+    auto col_heap =
+        reinterpret_cast<col_heap_element<ValueType, IndexType> *>(heap);
+
+    // first sweep: count nnz for each row
+#pragma omp parallel for
     for (size_type a_row = 0; a_row < num_rows; ++a_row) {
-        local_col_idxs.clear();
-        spgemm_insert_row(local_col_idxs, d, a_row);
-        spgemm_insert_row2(local_col_idxs, a, b, a_row);
-        c_row_ptrs[a_row] = local_col_idxs.size();
+        auto d_nz = d_row_ptrs[a_row];
+        auto d_end = d_row_ptrs[a_row + 1];
+        auto d_col = checked_load(d_cols, d_nz, d_end, sentinel);
+        c_row_ptrs[a_row] = spgemm_multiway_merge(
+            a_row, a, b, col_heap, [](size_type row) { return IndexType{}; },
+            [](ValueType, IndexType, IndexType &) {},
+            [&](IndexType col, IndexType &nnz) {
+                // skip smaller elements from d
+                while (d_col <= col) {
+                    d_nz++;
+                    nnz += d_col != col;
+                    d_col = checked_load(d_cols, d_nz, d_end, sentinel);
+                }
+                nnz++;
+            });
+        // handle the remaining columns from d
+        c_row_ptrs[a_row] += d_end - d_nz;
     }
 
     // build row pointers
@@ -293,18 +385,53 @@ void advanced_spgemm(std::shared_ptr<const OmpExecutor> exec,
     auto c_col_idxs = c_col_idxs_array.get_data();
     auto c_vals = c_vals_array.get_data();
 
-    map<IndexType, ValueType> local_row_nzs(exec);
-#pragma omp parallel for firstprivate(local_row_nzs)
+#pragma omp parallel for
     for (size_type a_row = 0; a_row < num_rows; ++a_row) {
-        local_row_nzs.clear();
-        spgemm_accumulate_row(local_row_nzs, d, vbeta, a_row);
-        spgemm_accumulate_row2(local_row_nzs, a, b, valpha, a_row);
-        // store result
-        auto c_nz = c_row_ptrs[a_row];
-        for (auto pair : local_row_nzs) {
-            c_col_idxs[c_nz] = pair.first;
-            c_vals[c_nz] = pair.second;
-            ++c_nz;
+        auto d_nz = d_row_ptrs[a_row];
+        auto d_end = d_row_ptrs[a_row + 1];
+        auto d_col = checked_load(d_cols, d_nz, d_end, sentinel);
+        auto d_val = checked_load(d_vals, d_nz, d_end, zero<ValueType>());
+        auto c_nz =
+            spgemm_multiway_merge(
+                a_row, a, b, heap,
+                [&](size_type row) {
+                    return std::make_pair(zero<ValueType>(), c_row_ptrs[row]);
+                },
+                [](ValueType val, IndexType,
+                   std::pair<ValueType, IndexType> &state) {
+                    state.first += val;
+                },
+                [&](IndexType col, std::pair<ValueType, IndexType> &state) {
+                    // handle smaller elements from d
+                    ValueType part_d_val{};
+                    while (d_col <= col) {
+                        if (d_col == col) {
+                            part_d_val = d_val;
+                        } else {
+                            c_col_idxs[state.second] = d_col;
+                            c_vals[state.second] = vbeta * d_val;
+                            state.second++;
+                        }
+                        d_nz++;
+                        d_col = checked_load(d_cols, d_nz, d_end, sentinel);
+                        d_val = checked_load(d_vals, d_nz, d_end,
+                                             zero<ValueType>());
+                    }
+                    c_col_idxs[state.second] = col;
+                    c_vals[state.second] =
+                        vbeta * part_d_val + valpha * state.first;
+                    state.first = zero<ValueType>();
+                    state.second++;
+                })
+                .second;
+        // handle remaining elements from d
+        while (d_col < sentinel) {
+            c_col_idxs[c_nz] = d_col;
+            c_vals[c_nz] = vbeta * d_val;
+            c_nz++;
+            d_nz++;
+            d_col = checked_load(d_cols, d_nz, d_end, sentinel);
+            d_val = checked_load(d_vals, d_nz, d_end, zero<ValueType>());
         }
     }
 }

--- a/omp/matrix/csr_kernels.cpp
+++ b/omp/matrix/csr_kernels.cpp
@@ -135,6 +135,15 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
 namespace {
 
 
+/**
+ * @internal
+ *
+ * Entry in a heap storing a column index and associated non-zero index
+ * (and row end) from a matrix.
+ *
+ * @tparam ValueType  The value type for matrices.
+ * @tparam IndexType  The index type for matrices.
+ */
 template <typename ValueType, typename IndexType>
 struct col_heap_element {
     using value_type = ValueType;
@@ -153,6 +162,15 @@ struct col_heap_element {
 };
 
 
+/**
+ * @internal
+ *
+ * Entry in a heap storing a entry (value and column index) and associated
+ * non-zero index (and row end) from a matrix.
+ *
+ * @tparam ValueType  The value type for matrices.
+ * @tparam IndexType  The index type for matrices.
+ */
 template <typename ValueType, typename IndexType>
 struct val_heap_element {
     using value_type = ValueType;
@@ -172,6 +190,20 @@ struct val_heap_element {
 };
 
 
+/**
+ * @internal
+ *
+ * Restores the binary heap condition downwards from a given index.
+ *
+ * The heap condition is: col(child) >= col(parent)
+ *
+ * @param heap  a pointer to the array containing the heap elements.
+ * @param idx  the index of the starting heap node that potentially
+ *             violates the heap condition.
+ * @param size  the number of elements in the heap.
+ * @tparam HeapElement  the element type in the heap. See col_heap_element and
+ *                      val_heap_element
+ */
 template <typename HeapElement>
 void sift_down(HeapElement *heap, typename HeapElement::index_type idx,
                typename HeapElement::index_type size)
@@ -193,6 +225,36 @@ void sift_down(HeapElement *heap, typename HeapElement::index_type idx,
 }
 
 
+/**
+ * @internal
+ *
+ * Generic SpGEMM implementation for a single output row of A * B using binary
+ * heap-based multiway merging.
+ *
+ * @param row  The row for which to compute the SpGEMM
+ * @param a  The input matrix A
+ * @param b  The input matrix B (its column indices must be sorted within each
+ *           row!)
+ * @param heap  The heap to use for this implementation. It must have as many
+ *              entries as the input row has non-zeros.
+ * @param init_cb  function to initialize the state for a single row. Its return
+ *                 value will be updated by subsequent calls of other callbacks,
+ *                 and then returned by this function. Its signature must be
+ *                 compatible with `return_type state = init_cb(row)`.
+ * @param step_cb  function that will be called for each accumulation from an
+ *                 entry of B into the output state. Its signature must be
+ *                 compatible with `step_cb(value, column, state)`.
+ * @param col_cb  function that will be called once for each output column after
+ *                all accumulations into it are completed. Its signature must be
+ *                compatible with `col_cb(column, state)`.
+ * @return the value initializaed by init_cb and updated by step_cb and col_cb
+ *
+ * @tparam HeapElement  the heap element type. See col_heap_element and
+ *                      val_heap_element
+ * @tparam InitCallback  functor type for init_cb
+ * @tparam StepCallback  functor type for step_cb
+ * @tparam ColCallback  functor type for col_cb
+ */
 template <typename HeapElement, typename InitCallback, typename StepCallback,
           typename ColCallback>
 auto spgemm_multiway_merge(size_type row,

--- a/omp/matrix/csr_kernels.cpp
+++ b/omp/matrix/csr_kernels.cpp
@@ -165,7 +165,7 @@ struct col_heap_element {
 /**
  * @internal
  *
- * Entry in a heap storing a entry (value and column index) and associated
+ * Entry in a heap storing an entry (value and column index) and associated
  * non-zero index (and row end) from a matrix.
  *
  * @tparam ValueType  The value type for matrices.
@@ -247,7 +247,9 @@ void sift_down(HeapElement *heap, typename HeapElement::index_type idx,
  * @param col_cb  function that will be called once for each output column after
  *                all accumulations into it are completed. Its signature must be
  *                compatible with `col_cb(column, state)`.
- * @return the value initializaed by init_cb and updated by step_cb and col_cb
+ * @return the value initialized by init_cb and updated by step_cb and col_cb
+ * @note If the columns of B are not sorted, the output may have duplicate
+ *       column entries.
  *
  * @tparam HeapElement  the heap element type. See col_heap_element and
  *                      val_heap_element

--- a/omp/preconditioner/jacobi_kernels.cpp
+++ b/omp/preconditioner/jacobi_kernels.cpp
@@ -326,28 +326,27 @@ inline bool invert_block(IndexType block_size, IndexType *perm,
 
 
 template <typename ReducedType, typename ValueType, typename IndexType>
-inline bool validate_precision_reduction_feasibility(
-    std::shared_ptr<const OmpExecutor> exec, IndexType block_size,
-    const ValueType *block, size_type stride)
+inline bool validate_precision_reduction_feasibility(IndexType block_size,
+                                                     const ValueType *block,
+                                                     size_type stride,
+                                                     ValueType *tmp_buffer,
+                                                     IndexType *tmp_perm)
 {
     using gko::detail::float_traits;
-    vector<ValueType> tmp(block_size * block_size, {}, exec);
-    vector<IndexType> perm(block_size, {}, exec);
-    std::iota(begin(perm), end(perm), IndexType{0});
+    std::iota(tmp_perm, tmp_perm + block_size, IndexType{0});
     for (IndexType i = 0; i < block_size; ++i) {
         for (IndexType j = 0; j < block_size; ++j) {
-            tmp[i * block_size + j] = static_cast<ValueType>(
+            tmp_buffer[i * block_size + j] = static_cast<ValueType>(
                 static_cast<ReducedType>(block[i * stride + j]));
         }
     }
     auto cond =
-        compute_inf_norm(block_size, block_size, tmp.data(), block_size);
-    auto succeeded =
-        invert_block(block_size, perm.data(), tmp.data(), block_size);
+        compute_inf_norm(block_size, block_size, tmp_buffer, block_size);
+    auto succeeded = invert_block(block_size, tmp_perm, tmp_buffer, block_size);
     if (!succeeded) {
         return false;
     }
-    cond *= compute_inf_norm(block_size, block_size, tmp.data(), block_size);
+    cond *= compute_inf_norm(block_size, block_size, tmp_buffer, block_size);
     return cond >= 1.0 &&
            cond * float_traits<remove_complex<ValueType>>::eps < 1e-3;
 }
@@ -371,52 +370,65 @@ void generate(std::shared_ptr<const OmpExecutor> exec,
     const auto prec = block_precisions.get_data();
     const auto group_size = storage_scheme.get_group_size();
     const auto cond = conditioning.get_data();
+    const auto num_threads = omp_get_max_threads();
+    // group_size blocks for the actual blocks, 1 for temporary storage
+    const auto parallel_blocks = (group_size + 1) * num_threads;
+    Array<ValueType> blocks_storage{
+        exec, static_cast<size_type>(parallel_blocks * max_block_size *
+                                     max_block_size)};
+    Array<IndexType> perm_storage{
+        exec, static_cast<size_type>(parallel_blocks * max_block_size)};
+    Array<uint32> pr_descriptor_storage(exec, parallel_blocks);
 #pragma omp parallel for
     for (size_type g = 0; g < num_blocks; g += group_size) {
-        vector<Array<ValueType>> block(group_size, {}, exec);
-        vector<Array<IndexType>> perm(group_size, {}, exec);
-        vector<uint32> pr_descriptors(group_size, uint32{} - 1, exec);
+        const auto thread_id = omp_get_thread_num();
+        const auto thread_offset = thread_id * (group_size + 1);
+        auto local_blocks_tmp =
+            blocks_storage.get_data() +
+            (thread_offset * max_block_size * max_block_size);
+        auto local_blocks = local_blocks_tmp + max_block_size * max_block_size;
+        auto local_perms_tmp =
+            perm_storage.get_data() + (thread_offset * max_block_size);
+        auto local_perms = local_perms_tmp + max_block_size;
+        auto pr_descriptors = pr_descriptor_storage.get_data() + thread_offset;
+        std::fill_n(pr_descriptors, group_size, uint32{} - 1);
         // extract group of blocks, invert them, figure out storage precision
         for (size_type b = 0; b < group_size; ++b) {
             if (b + g >= num_blocks) {
                 break;
             }
             const auto block_size = ptrs[g + b + 1] - ptrs[g + b];
-            block[b] = Array<ValueType>(exec, block_size * block_size);
-            perm[b] = Array<IndexType>(exec, block_size);
-            std::iota(perm[b].get_data(), perm[b].get_data() + block_size,
-                      IndexType(0));
-            extract_block(system_matrix, block_size, ptrs[g + b],
-                          block[b].get_data(), block_size);
+            auto block = &local_blocks[b * max_block_size * max_block_size];
+            auto perm = &local_perms[b * max_block_size];
+            std::iota(perm, perm + block_size, IndexType{0});
+            extract_block(system_matrix, block_size, ptrs[g + b], block,
+                          block_size);
             if (cond) {
                 cond[g + b] =
-                    compute_inf_norm(block_size, block_size,
-                                     block[b].get_const_data(), block_size);
+                    compute_inf_norm(block_size, block_size, block, block_size);
             }
-            invert_block(block_size, perm[b].get_data(), block[b].get_data(),
-                         block_size);
+            invert_block(block_size, perm, block, block_size);
             if (cond) {
                 cond[g + b] *=
-                    compute_inf_norm(block_size, block_size,
-                                     block[b].get_const_data(), block_size);
+                    compute_inf_norm(block_size, block_size, block, block_size);
             }
             const auto local_prec = prec ? prec[g + b] : precision_reduction();
             if (local_prec == precision_reduction::autodetect() && cond) {
                 using preconditioner::detail::get_supported_storage_reductions;
                 pr_descriptors[b] = get_supported_storage_reductions<ValueType>(
                     accuracy, cond[g + b],
-                    [&exec, &block_size, &block, &b] {
+                    [&] {
                         using target = reduce_precision<ValueType>;
                         return validate_precision_reduction_feasibility<target>(
-                            exec, block_size, block[b].get_const_data(),
-                            block_size);
+                            block_size, block, block_size, local_blocks_tmp,
+                            local_perms_tmp);
                     },
-                    [&exec, &block_size, &block, &b] {
+                    [&] {
                         using target =
                             reduce_precision<reduce_precision<ValueType>>;
                         return validate_precision_reduction_feasibility<target>(
-                            exec, block_size, block[b].get_const_data(),
-                            block_size);
+                            block_size, block, block_size, local_blocks_tmp,
+                            local_perms_tmp);
                     });
             } else {
                 pr_descriptors[b] = preconditioner::detail::
@@ -426,7 +438,7 @@ void generate(std::shared_ptr<const OmpExecutor> exec,
 
         // make sure everyone in the group uses the same precision
         const auto p = preconditioner::detail::get_optimal_storage_reduction(
-            std::accumulate(begin(pr_descriptors), end(pr_descriptors),
+            std::accumulate(pr_descriptors, pr_descriptors + group_size,
                             uint32{} - 1,
                             [](uint32 x, uint32 y) { return x & y; }));
 
@@ -439,11 +451,12 @@ void generate(std::shared_ptr<const OmpExecutor> exec,
                 prec[g + b] = p;
             }
             const auto block_size = ptrs[g + b + 1] - ptrs[g + b];
+            auto block = &local_blocks[b * max_block_size * max_block_size];
+            auto perm = &local_perms[b * max_block_size];
             GKO_PRECONDITIONER_JACOBI_RESOLVE_PRECISION(
                 ValueType, p,
                 permute_and_transpose_block(
-                    block_size, perm[b].get_data(), block[b].get_data(),
-                    block_size,
+                    block_size, perm, block, block_size,
                     reinterpret_cast<resolved_precision *>(
                         blocks.get_data() +
                         storage_scheme.get_group_offset(g + b)) +

--- a/omp/test/matrix/csr_kernels.cpp
+++ b/omp/test/matrix/csr_kernels.cpp
@@ -91,12 +91,19 @@ protected:
 
     template <typename MtxType>
     std::unique_ptr<MtxType> gen_mtx(int num_rows, int num_cols,
-                                     int min_nnz_row)
+                                     int min_nnz_row, int max_nnz_row)
     {
         return gko::test::generate_random_matrix<MtxType>(
             num_rows, num_cols,
-            std::uniform_int_distribution<>(min_nnz_row, num_cols),
+            std::uniform_int_distribution<>(min_nnz_row, max_nnz_row),
             std::normal_distribution<>(-1.0, 1.0), rand_engine, ref);
+    }
+
+    template <typename MtxType>
+    std::unique_ptr<MtxType> gen_mtx(int num_rows, int num_cols,
+                                     int min_nnz_row)
+    {
+        return gen_mtx<MtxType>(num_rows, num_cols, min_nnz_row, num_cols);
     }
 
     void set_up_apply_data(int num_vectors = 1)
@@ -253,6 +260,60 @@ TEST_F(Csr, SimpleApplyToCsrMatrixIsEquivalentToRef)
 
     GKO_ASSERT_MTX_NEAR(square_dmtx, square_mtx, 1e-14);
     GKO_ASSERT_MTX_EQ_SPARSITY(square_dmtx, square_mtx);
+    ASSERT_TRUE(square_dmtx->is_sorted_by_column_index());
+}
+
+
+TEST_F(Csr, SimpleApplyToSparseCsrMatrixIsEquivalentToRef)
+{
+    set_up_apply_data();
+    auto mtx2 =
+        gen_mtx<Mtx>(mtx->get_size()[1], square_mtx->get_size()[1], 0, 10);
+    auto dmtx2 = Mtx::create(omp, mtx2->get_size());
+    dmtx2->copy_from(mtx2.get());
+
+    mtx->apply(mtx2.get(), square_mtx.get());
+    dmtx->apply(dmtx2.get(), square_dmtx.get());
+
+    GKO_ASSERT_MTX_EQ_SPARSITY(square_dmtx, square_mtx);
+    GKO_ASSERT_MTX_NEAR(square_dmtx, square_mtx, 1e-14);
+    ASSERT_TRUE(square_dmtx->is_sorted_by_column_index());
+}
+
+
+TEST_F(Csr, SimpleApplySparseToSparseCsrMatrixIsEquivalentToRef)
+{
+    set_up_apply_data();
+    auto mtx1 = gen_mtx<Mtx>(mtx->get_size()[0], mtx->get_size()[1], 0, 10);
+    auto mtx2 =
+        gen_mtx<Mtx>(mtx->get_size()[1], square_mtx->get_size()[1], 0, 10);
+    auto dmtx1 = Mtx::create(omp, mtx1->get_size());
+    auto dmtx2 = Mtx::create(omp, mtx2->get_size());
+    dmtx1->copy_from(mtx1.get());
+    dmtx2->copy_from(mtx2.get());
+
+    mtx1->apply(mtx2.get(), square_mtx.get());
+    dmtx1->apply(dmtx2.get(), square_dmtx.get());
+
+    GKO_ASSERT_MTX_EQ_SPARSITY(square_dmtx, square_mtx);
+    GKO_ASSERT_MTX_NEAR(square_dmtx, square_mtx, 1e-14);
+    ASSERT_TRUE(square_dmtx->is_sorted_by_column_index());
+}
+
+
+TEST_F(Csr, SimpleApplyToEmptyCsrMatrixIsEquivalentToRef)
+{
+    set_up_apply_data();
+    auto mtx2 =
+        gen_mtx<Mtx>(mtx->get_size()[1], square_mtx->get_size()[1], 0, 0);
+    auto dmtx2 = Mtx::create(omp, mtx2->get_size());
+    dmtx2->copy_from(mtx2.get());
+
+    mtx->apply(mtx2.get(), square_mtx.get());
+    dmtx->apply(dmtx2.get(), square_dmtx.get());
+
+    GKO_ASSERT_MTX_EQ_SPARSITY(square_dmtx, square_mtx);
+    GKO_ASSERT_MTX_NEAR(square_dmtx, square_mtx, 1e-14);
     ASSERT_TRUE(square_dmtx->is_sorted_by_column_index());
 }
 

